### DISCUSSION TITLE
Is there a general way to process an endpoint response (opposite of dependency)

### DISCUSSION THREAD

**Discussion by rossmacarthur** [fastapi/fastapi#11155] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from pydantic import BaseModel

app = FastAPI()

class Item(BaseModel):
    id: int

@app.get("/items/{item_id}")
def read_item(item_id: int) -> Item:
    return Item(id=item_id)
```

### Description

I have a project with many endpoints defined like in the above code, where the response is a pydantic model. I would like to post process these responses _before_ the pydantic type is dumped. 

I also need access to app state and other dependencies in this context. Is this possible? 

I looked at the following options and they didn't seem suitable:

- **Custom response class**: this receives the response model after it has been converted to python objects
- **Custom route class**: similar problem to above
- **Middleware**: similar problem to above, the response is already serialized
- **Patching `jsonable_encoder`**: I don't have access to app state or dependencies here
- **Patching `serialize_response`**: Same as above

Is there an option I'm missing, besides applying a decorator to every single endpoint method? Is it possible to modify `APIRoute.dependant` somehow?

### Operating System

Linux

### FastAPI Version

0.108.0

### Pydantic Version

2.6.1

### Python Version

Python 3.11.7

**Comment by sudoAlireza** [comment]
I'm not sure that I understand your question but I think it's a Pydantic question. you can look at `model_dump(...)` in Pydantic Documentation [Serialization](https://docs.pydantic.dev/latest/concepts/serialization/).
If it wasn't helpful, please provide a clear example of what you intend to do.

**Reply by rossmacarthur** [reply]
I can't use `model_dump(..)` because I need access to fastapi dependencies. Basically I want to lazily fill certain data in the response before returning and serializing to JSON, in my specific context I need a redis connection which is typically provided as a fastapi dependency for the endpoint.

Here is an example of how I am currently doing it, which basically applies a decorator to everything.


```python

class CustomRoute(APIRoute):
    def __init__(self, path: str, endpoint: Callable[..., Any], **kwargs: Any) -> None:
        super().__init__(path, endpoint, **kwargs)
        maybe_rewrite_dependent_call(path, self.dependant, self.response_model)


def maybe_rewrite_dependent_call(path: str, dep: Dependant, response_model: Any) -> None:
    if "redis" not in {d.name for d in dep.dependencies}:
        raise RuntimeError(f"{path}: redis not found in dependencies")
    if dep.call is None:
        raise RuntimeError(f"{path}: dependant call is None")
    if not asyncio.iscoroutinefunction(dep.call):
        raise RuntimeError(f"{path}: dependant call is not a coroutine function")
    dep.call = _new_dependant_call(dep.call)


def _new_dependant_call(endpoint: Endpoint) -> Endpoint:
    @wraps(endpoint)
    async def _rewrite(*args: Any, **kwargs: Any) -> Any:
        response = await endpoint(*args, **kwargs)
        await _lazy_fill_data(kwargs["redis"], response)
        return response
    return _rewrite

async def _lazy_fill_data(redis: Redis, response: Any) -> None:
    ...
````
